### PR TITLE
Add direct s3 support to the streaming dataset

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds a mechanism to read directly from s3 and generate a cache folder in the home of the user.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--19044.org.readthedocs.build/en/19044/

cc @borda

### CODE REVIEW

File: tests/tests_data/streaming/test_dataset.py [hunk 1]
```diff
@@ -307,7 +296,7 @@ def test_dataset_cache_recreation(tmpdir):
 
 def test_try_create_cache_dir():
     with mock.patch.dict(os.environ, {}, clear=True):
-        assert _try_create_cache_dir("any") is None
+        assert f"{os.sep}".join(["chunks", "100b8cad7cf2a56f6df78f171f97a1ec", "0"]) in _try_create_cache_dir("any")
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_dataset.py:299]: Can you use `os.path.join()`?

File: src/lightning/data/streaming/dataset.py [hunk 2]
```diff
@@ -194,9 +208,13 @@ def __next__(self) -> Any:
 
 
 def _try_create_cache_dir(input_dir: str, shard_rank: int = 0) -> Optional[str]:
-    if "LIGHTNING_CLUSTER_ID" not in os.environ or "LIGHTNING_CLOUD_PROJECT_ID" not in os.environ:
-        return None
     hash_object = hashlib.md5(input_dir.encode())
+    if "LIGHTNING_CLUSTER_ID" not in os.environ or "LIGHTNING_CLOUD_PROJECT_ID" not in os.environ:
+        cache_dir = os.path.join(
+            os.path.expanduser("~"), ".lightning", "chunks", hash_object.hexdigest(), str(shard_rank)
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:214]: we could define the chunks base folder in constants.py

and using `Path.home()` would be nice.

**Review by awaelchli — APPROVED** [review]

File: src/lightning/data/streaming/dataset.py [hunk 3]
```diff
@@ -26,16 +27,25 @@
 from lightning.data.streaming.shuffle import FullShuffle, NoShuffle, Shuffle
 from lightning.data.utilities.env import Environment, _DistributedEnv, _WorkerEnv
 
+
+@dataclass
+class RemoteDir:
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:32]: Let's define this at the bottom. It's nicer to have the core api definition (streaming dataset) at the top of the file. Then you'll need to use string references, e.g. `input_dir: Union[str, "RemoteDir"],` for types.

File: tests/tests_data/streaming/test_dataset.py (hunk 1 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py:299]: done !

File: src/lightning/data/streaming/dataset.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:214]: done !

File: src/lightning/data/streaming/dataset.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:32]: Sure.

File: .github/workflows/ci-tests-data.yml [hunk 4]
```diff
@@ -94,7 +94,12 @@ jobs:
 
       - name: Testing Data
         working-directory: tests/tests_data
+        env:
```
Comment (Borda) [inline .github/workflows/ci-tests-data.yml:97]: Note that this means no developer out of lightning org can contribute

File: .github/workflows/ci-tests-data.yml [hunk 5]
```diff
@@ -94,7 +94,12 @@ jobs:
...
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_PUB_ACCESS_KEY }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_PUB_SECRET_KEY }}
+          AWS_DEFAULT_REGION: ${{ secrets.AWS_REGION }}
```
Comment (Borda) [inline .github/workflows/ci-tests-data.yml:100]: What is the user and permissions? 
Pls, reach me for a validation vulnerability check

**Review by Borda — CHANGES_REQUESTED** [review]
until the AWS credentials are clatified

File: .github/workflows/ci-tests-data.yml (hunk 4 above)
Comment (tchaton) [reply .github/workflows/ci-tests-data.yml:97]: The bucket is public
Just need to be auth to AWS
Any dev can contribute if they use their own credentials

File: .github/workflows/ci-tests-data.yml (hunk 5 above)
Comment (tchaton) [reply .github/workflows/ci-tests-data.yml:100]: The bucket is public
Just need to be auth to AWS
Any dev can contribute if they use their own credentials

File: .github/workflows/ci-tests-data.yml (hunk 4 above)
Comment (tchaton) [reply .github/workflows/ci-tests-data.yml]: removed.

File: .github/workflows/ci-tests-data.yml (hunk 5 above)
Comment (tchaton) [reply .github/workflows/ci-tests-data.yml]: removed.

**Review by lantiga — APPROVED** [review]
LGTM!

**Pull request merged**